### ISSUE TITLE
Bezier curve drawing tool

### ISSUE THREAD

**Issue by gammalogic** [Intervention/image#1335]
**Describe the feature you'd like**
Provide a bezier curve drawing tool that can work in GD, similar to the pathCurveToQuadraticBezierAbsolute() and pathCurveToRelative() methods provided by Imagick.

**Is your feature request related to a problem? Please describe.**
Lack of bezier curve drawing support in GD means that a 3rd party library needs to implement it in some form.

GD is so prevalent on PHP servers that this kind of tool (however basic) would be really useful. I need a bezier curve drawing tool in an application I'm working on at the moment (which uses the GD extension) and I eventually found some code in an old Stack Overflow post that does what I want, but not everyone is going to have the patience or time to find example code and get it working. I think it would add value to II to include support for it. It appears that GD is still in active development and I found a reference to a DrawBezier function in the source code, but support doesn't seem to have been extended to the PHP implementation yet. I have received good results from the code I found, which can draw an accurate cubic bezier curve:

![Cubic bezier curve drawn using the GD extension](https://github.com/Intervention/image/assets/28597602/3b05dba1-dff6-4214-9ed1-ee3f21f308d9)

**Comment by olivervogel** [comment]
This sounds like an interesting idea.

Intervention Image is a standardized API for GD and Imagick. The main goal is to make the results of the API independent of the driver used and thus ensure interchangeability.

Of course, this goal is not always 100% achievable. However, I have always done my best to get as close to 100% as possible. The same should happen with this feature.

### 1. Bezier Curve Drawing with GD

It must first be clarified whether it is possible to draw Bezier curves with GD and if so, what type, as there are different types such as cubic, quadratic, etc.

Have you already gained experience here and can you perhaps even show a code example?

### 2. Design of Public API

In the second step, a public API should be designed with which this curve drawing can be executed. Of course, this is also dependent on the possibilities and Bezier types that are possible with Imagick.

Unfortunately, I'm not really sure which features need to be taken into account here.

A simple call could look like this:

```php
// draw bezier curve
$image->drawBezierCurve(function (BezierCurveFactory $curve) {
    $curve->color('ff5500');
    $curve->point(10, 10);
    $curve->point(150, 150);
    $curve->point(40, 180);
    $curve->point(60, 100);
});
```

Although, as I said, I'm not sure whether all the necessary things have been taken into account here. For example, there are also control points for Bezier curves, which also have to be mapped somehow.

Can you say more about this?

### 3. Implementation

Once these questions 1 and 2 have been answered, it's time for the implementation.

**Comment by gammalogic** [comment]
This is a very basic animation that I've put together that shows how the points in a cubic bezier curve can be calculated and drawn with a line tool:

![Cubic bezier curve calculation animatin](https://github.com/Intervention/image/assets/28597602/8aeb2607-f4ec-48b1-89bb-bc812c95d6d8)

While researching this I found a very helpful [interactive guide](https://www.summbit.com/blog/bezier-curve-guide/#cubic-bezier-curve-construction) that explains how the cubic is split to calculate the curve; to replicate the steps in my animation you just need to set the sliders to halfway to replicate the **t** (ratio) value of 0.5. The maths involved is fairly simple from a computational perspective (nothing more complex than additions and multiplications) and it took approximately 30 splitting operations to draw the curve shown in my animation. The recursion depth was set to 5 but a higher value would produce better results.

**Comment by gammalogic** [comment]
This is a comparison of the output from the `Imagick::pathCurveToAbsolute()` function and my custom GD function:

![Cubic Bezier Curve Comparison with Imagick and GD](https://github.com/Intervention/image/assets/28597602/bda95bff-0c22-4276-92b9-42b6a615f7c3)

Allowing for the difference in quality due to Imagick using anti-aliasing and GD not using anti-aliasing, the output from GD is good for 1 pixel wide, acceptable up to 5 pixels wide, and artifacts start appearing after that point (the right-hand curves are 20 pixels wide). These artifacts are in part due to GD's `imageline()` function not drawing the start and end caps of the line segments perpendicularly. As an experiment, I updated my custom GD function to over-draw an ellipse with a diameter equivalent to the line segment width at each point and this partly solves the problem:

![Cubic Bezier Curve Comparison GD Ellipse plus Line](https://github.com/Intervention/image/assets/28597602/14410066-b6bf-4f34-a643-2236abe7fcb6)

I'm not sure if the output is good enough for inclusion in II, but given GD's limitations I think the results are certainly acceptable and (personally) I would rather have the tool than not.

**Comment by olivervogel** [comment]
I think these are great results! Even more so when you consider that GD doesn't actually include the function.

The differences with anti-aliasing can also be observed with the other existing draw functions. I don't think this is a deal breaker, as GD is often very limited anyway. 

As I wrote before, the goal should be a 100% match across drivers. As already noted, this is almost impossible to achieve in reality. However, I always try to get as close as possible and I think this is already done here.

I even think that limiting the line width to a maximum of 5 pixels would be an acceptable option to avoid artifacts. There was a similar problem with the implementation of the [text outline function](https://image.intervention.io/v3/modifying/text-fonts#text-stroke-outline) (with GD of course). Again, the [stroke width was limited to 10 pixels](https://github.com/Intervention/image/blob/develop/src/Typography/Font.php#L155).

**Comment by gammalogic** [comment]
Thanks for your feedback. Your suggestion of limiting the line width is a good compromise. I am still trying to get the calculations finished as I am rewriting them to avoid re-using any of the sample code then, if everything is ok, I will attempt a PR.

**Comment by gammalogic** [comment]
Just a quick update: I have now implemented the bezier curve calculations using the Bernstein polynomial form, which is computationally not expensive. I have also found a good solution to the line thickness issue, by calculating inside/outside offset values for each line segment and then connecting the line segments together and drawing the curve using a filled polygon:

![Cubic Bezier Curve Comparison GD Filled Polygon](https://github.com/Intervention/image/assets/28597602/c4b562d4-dbd2-453f-84b3-cbb3adec8de4)

This also solves the problem of the incorrect start/end cap angles, and the curve now looks much closer to how it would look using Imagick; still. For line thicknesses of 1 pixel the curve would still be drawn using line segments.

I have just started looking at the Intervention Image source code and will attempt a PR soon.

**Comment by olivervogel** [comment]
> I have just started looking at the Intervention Image source code and will attempt a PR soon.

That's great news. I'm looking forward to your PR. If there's anything I can do to help, please let me know.

**Comment by olivervogel** [comment]
A brief follow-up on the modifier architecture.

1. The basis is a generic modifier class whose object can be passed in `ImageInterface::modify()`. A close example would be [`Intervention\Image\Modifiers\DrawLineModifer::class`](https://github.com/Intervention/image/blob/develop/src/Modifiers/DrawLineModifier.php).
2. The respective modifier classes of the individual drivers are inherited from this generic class. These are found and created internally by using the same class name (I call this process "spezializing"). Then the class execute the respective modifying actions using the driver's resources. For GD and the DrawLineModifer example, this would be [`Intervention\Image\Drivers\Gd\Modifiers\DrawLineModifer::class.`](https://github.com/Intervention/image/blob/develop/src/Drivers/Gd/Modifiers/DrawLineModifier.php) 

As I said, if you have any questions or I can help, let me know.

**Comment by gammalogic** [comment]
Thank you. I have added the bezier curve calculation code to the GD driver script because these calculations will not be needed by Imagick. I have also added an exception in that script to generate a fatal error if the user does not specify exactly 4 points when creating the curve.

To create a new `Bezier` class, I have modified the `Polygon` class. In that class, I can see that there are functions to offset the x/y coordinates, rotate and align the polygon, but I could not see how to access this functionality.

For example, I tried this:

```php
$image->drawPolygon(function (PolygonFactory $polygon) {
    $polygon->point(100, 100); // add point of polygon
    $polygon->point(150, 150); // add point
    $polygon->point(40, 180); // add point
    $polygon->point(60, 100); // add point
    $polygon->background('#b35187'); // background color
    $polygon->border('#ff0', 6); // border color and border width
    $polygon->movePointsX(100); // move all x ordinates 100 pixels
});
```

but movePointsX() cannot be called directly because it is not defined in the `PolygonFactory` class. Is there a correct way to access it from a user script? As the bezier curve is (initially) a 4 point polygon, these transformations could be used for it as well.

**Comment by olivervogel** [comment]
> but movePointsX() cannot be called directly because it is not defined in the PolygonFactory class.

I don't think it is absolutely necessary to have access to this method. The position is already defined by specifying the points. Or is there a special case where it is indispensable?

What I have already thought about is to be able to pass the geometric object directly in addition to the callable type.

Something like this:

```php
$image->drawPolygon(new Polygon());
```
However, this issue is about bezier curves I have created a [new issue](https://github.com/Intervention/image/issues/1358) for this.

**Comment by gammalogic** [comment]
I did not know if these functions are meant to be accessed by the user? They might be useful if the user has a complex polygon (or bezier) that needs to be rotated; the x/y move operations are possibly less important. To rotate a polygon, a pass-through function could be added to the `PolygonFactory` class:

```php
public function rotate(float $angle): self
{
    $this->polygon->rotate($angle);

    return $this;
}
```
and if I then draw the polygon using the code below the rotation works:

```php
$image->drawPolygon(function (PolygonFactory $polygon) {
    $polygon->point(100, 100); // add point of polygon
    $polygon->point(150, 150); // add point
    $polygon->point(40, 180); // add point
    $polygon->point(60, 100); // add point
    $polygon->background('#b35187'); // background color
    $polygon->border('#ff0', 6); // border color and border width
    $polygon->rotate(-45); // rotate polygon by -45 degrees
});
```

I wanted to ask about these functions just because I could not see where or how they are used.

**Comment by olivervogel** [comment]
> I wanted to ask about these functions just because I could not see where or how they are used.

Not all functions are currently being used, but this may well change in the future. For now, let's ignore these functions for the Bezier Curves feature.

**Comment by gammalogic** [comment]
> Not all functions are currently being used, but this may well change in the future. For now, let's ignore these functions for the Bezier Curves feature.

Ok, I will ignore them.

I have implemented most of the functionality now for both the Imagick and GD drivers. This image shows the output from Imagick (left) and GD (right):

![II Bezier Curve Comparison](https://github.com/Intervention/image/assets/28597602/bd2343a3-8983-46b1-a0b6-fb554b417353)

The GD functionality works quite well apart from where the curve overlaps, which I will try and fix. It is also possible to generate both quadratic and cubic bezier curves; the type is automatically selected based on how many control points the user supplies (3 control points for quadratic, 4 control points for cubic).

**Comment by olivervogel** [comment]
Looks very promising.

**Comment by olivervogel** [comment]
This feature is available as of version [3.7.0](https://github.com/Intervention/image/releases/tag/3.7.0).
